### ISSUE TITLE
Remove `System.import()` usage in favor of `import()`

### ISSUE THREAD

**Issue by filipesilva** [angular/angular#21560]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

Currently Angular core makes use of a `System.import()` call to load templates (html/css) and lazy modules:

https://github.com/angular/angular/blob/3bcc0e6f765f86ce96dbb33edf4ba0cbc53b3322/packages/core/src/linker/system_js_ng_module_factory_loader.ts#L67-L70

In the past `System.import` was expected to become a standard, but `import()` was settled upon instead.

Modern tooling is beginning to deprecate (notably Webpack 4 https://github.com/webpack/webpack/pull/6321) `System.import()` detection in favor of the standard and it should not be unexpected that more follow suit.

## Expected behavior

The standard `import()` is used, so that tooling that supports code splitting can detect it. 

## Minimal reproduction of the problem with instructions
The SystemJS loader can be found at https://github.com/angular/angular/blob/master/packages/core/src/linker/system_js_ng_module_factory_loader.ts.

## What is the motivation / use case for changing the behavior?

Remain compatible with modern tooling.

## Environment

<pre><code>
Angular version: 2, 4 and 5
</code></pre>

/cc @hansl @bradlygreen @IgorMinar @robwormald

**Comment by robwormald** [comment]
In theory I'm all for this, but what happens with ES5 code? What does `import` transpile down to do for a UMD or CommonJS bundle? It doesn't seem like this is something we can do in a backwards compatible fashion, unless I'm missing something.

I'd prefer to drop this whole abstraction in favor of just doing it inline:
```ts
{ loadChildren: import('./some-child').then(mod => mod.SomeChild) } 
```
but that's going to be gated on inlining factories (so we don't need to do the rewrite), or making CLI smart enough to pick up imports like ^ and rewrite them.

**Comment by filipesilva** [comment]
Personally I'd prefer dropping the abstraction as well, but haven't considered the implications of that option much.

The CLI could be smart enough to pick up those imports I think. We already have to do that for `System.import`.

**Comment by Toxicable** [comment]
It was my understanding that `import()` was only available with when targeting `es2015` as a requirement of TS, meaning it does not support es5.

**Comment by filipesilva** [comment]
It's not a big problem for applications proper. It's more of an issue for libraries and packaging concerns. If Angular were to have `import()` in it's codebase, it could only be consumed by tooling that understood `import()` and know to also load what the statement was referencing.

On formats like UMD or CommonJS it would have to be provided externally. I don't think it transpiles down to anything.

**Comment by trotyl** [comment]
@Toxicable Dynamic import has no restriction on `target`, but on `module`. All module types except `es2015` are allowed, for example with source:

```typescript
export function load(path) {
    return import(path);
}
```

To `module: esnext`:

```javascript
export function load(path) {
    return import(path);
}
```

To `module: commonjs`:

```javascript
"use strict";
exports.__esModule = true;
function load(path) {
    return Promise.resolve().then(function () { return require(path); });
}
exports.load = load;
```

To `module: amd`:

```javascript
define(["require", "exports"], function (require, exports) {
    "use strict";
    exports.__esModule = true;
    function load(path) {
        return new Promise(function (resolve_1, reject_1) { require([path], resolve_1, reject_1); });
    }
    exports.load = load;
});
```

To `module: system`:

```javascript
System.register([], function (exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    function load(path) {
        return context_1["import"](path);
    }
    exports_1("load", load);
    return {
        setters: [],
        execute: function () {
        }
    };
});
```

To `module: umd`:

```javascript
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    var __syncRequire = typeof module === "object" && typeof module.exports === "object";
    exports.__esModule = true;
    function load(path) {
        return __syncRequire ? Promise.resolve().then(function () { return require(path); }) : new Promise(function (resolve_1, reject_1) { require([path], resolve_1, reject_1); });
    }
    exports.load = load;
});
```

**Comment by Toxicable** [comment]
@trotyl Right, my mistake I confused the two of them.
However, this would still be a breaking change since most people use `es2015` since that's the CLI's default `module`

**Comment by trotyl** [comment]
@Toxicable Just did a check, the `module` restriction is only meant for `.ts` source files, not for `.d.ts` declaration files.

The emited `.d.ts` is:

```typescript
export declare function load(path: string): Promise<any>;
```

There's no pattern indicating the dynamic import.

**Comment by Toxicable** [comment]
Oh that's an interesting point maybe it would be fine to make that change then

**Comment by filipesilva** [comment]
Great breakdown @trotyl!

**Comment by IgorMinar** [comment]
@filipesilva will this be still compatible with webpack 3?

**Comment by filipesilva** [comment]
@IgorMinar Webpack 3 supports `import()` as the primary code splitting construct https://webpack.js.org/guides/code-splitting/#dynamic-imports.

`System.import()` isn't listed in the docs even, but it is supported since the Webpack 3 betas.

**Comment by songawee** [comment]
FWIW, I `{ loadChildren: import('./some-cool-module').then(mod => mod.ModuleName) } ` and it seems to work in JIT mode, but I had to switch to `{ loadChildren: './path-to-module#ModuleName' }`  to get it to work with AOT.

Was using Angular 5.x.x and Webpack 3.x.x.

If `import` was used in Angular, would that mean that we could code split in more places than just the routes and still have AOT work? That would be amazing. Thinking the equivalent of something like the react-loadable project where component level splitting was an option without having to tie the split to the router.

**Comment by trotyl** [comment]
@songawee You're talking about #21522, which is not covered in this one. Using dynamic import for routes is already possible today (but requiring wrap to exported function declaration).

The path mapping is not in scope of this repo, but tooling specific (like Angular CLI).

**Comment by trotyl** [comment]
Suggested design:

+ Targeting `"module": "esnext"` in `tsconfig.json` to let it emitted as is;
+ Passing through the `import()` syntax to `es` and `esm` builds by rollup (https://github.com/rollup/rollup/pull/1985);
+ Transforming to `SystemJS.import()` and fallback to `System.import()` in `umd` builds by manual replacement step (https://github.com/rollup/rollup-plugin-replace);
+ Deprecate the `System.import()` fallback in `umd` builds;

This way should be non-breaking for Webpack, Rollup and UMD users, not quite sure about Closure Compiler.

**Comment by everedifice** [comment]
Guys. If the `System.import` deprecated. how can i import css files?

When i trying to import css file by `import().then` compiler throw error like

```
ERROR in src/app/app.component.ts(18,12): error TS2307: Cannot find module '../assets/reset.css'.
```

**Comment by thw0rted** [comment]
@everedifice this isn't really the place to be asking about the mechanics of `import()`, but the answer is, it depends.  Does your environment (browser / Node) support `import()` for text assets directly?  Does your build environment (e.g. webpack) handle it through a loader?  Also, are you actually trying to process the raw text of your CSS in your JS or do you just want to add a stylesheet to the page?  You might be able to add your CSS to your build configuration instead of using `import()`.

**Comment by everedifice** [comment]
@thw0rted Thank you for your reply. I just sad deprecation of the feature that load css file dynamically. Actually i expected comments like you said.

I realize just few minutes before. is there way of dynamic loading of module without registration of router by using `NgModuleFactoryLoader`. I load css resources by `require()` in lazyload modules. It is work just i expected.

**Comment by kevinmims** [comment]
@filipesilva has there been any progress on this issue?

**Comment by jlguenego** [comment]
For info, webpack can disable filtered warning with its stats options.
For instance you can add to your webpack config:
`warningsFilter: /System.import/,`

See https://webpack.js.org/configuration/stats/ for more info.

**Comment by jerguslejko** [comment]
any updates?

**Comment by iantrudell** [comment]
Another month passed, any progress?

**Comment by Andrey-Pavlov** [comment]
@jlguenego  https://github.com/angular/angular/issues/21560#issuecomment-406342840
Works for webpack console and doesn't for browser console.
For **webpack-dev-server** this property needs to be in the devServer object.
Still need a solid solution or workaround.

**Comment by Moximillian** [comment]
As a workaround, used [webpack-filter-warnings-plugin](https://www.npmjs.com/package/webpack-filter-warnings-plugin) to hide the warning from browser.

```
const FilterWarningsPlugin = require('webpack-filter-warnings-plugin');

module.exports = {
  plugins: [
    new FilterWarningsPlugin({ 
      exclude: /System.import/ 
    })
  ]
};
```

**Comment by Benno007** [comment]
> For info, webpack can disable filtered warning with its stats options.
> For instance you can add to your webpack config:
> `warningsFilter: /System.import/,`
> 
> See https://webpack.js.org/configuration/stats/ for more info.

Didn't work for me, I had to add --client-log-level error to our webpack-dev-server cli arguments to prevent it from showing in the browser. We also use HMR. Still seeing `compiled with warnings` in the command line, but don't actually see any warnings in the terminal...

**Comment by kuhnroyal** [comment]
The `webpack-filter-warnings-plugin` works as temporary workaround, just put it before `friendly-errors-webpack-plugin` and `simple-progress-webpack-plugin`, if you use those.

**Comment by bitterloa** [comment]
> > For info, webpack can disable filtered warning with its stats options.
> > For instance you can add to your webpack config:
> > `warningsFilter: /System.import/,`
> > See https://webpack.js.org/configuration/stats/ for more info.
> 
> Didn't work for me, I had to add --client-log-level error to our webpack-dev-server cli arguments to prevent it from showing in the browser. We also use HMR. Still seeing `compiled with warnings` in the command line, but don't actually see any warnings in the terminal...

Just wanted to mention that using webpack 4's built in `warningsFilter: /System.import/,` did *not* work for me either. And I put this both in the main config and also in the `devServer > stats` config and no dice, I was still seeing the errors. 

The thing that did work was using the `FilterWarningsPlugin` as mentioned before by @Moximillian

**Comment by virgil-av** [comment]
The way I found it to work in Angular 7:

```
TranslateModule.forRoot({
      loader: {
        provide: TranslateLoader,
        useClass: TranslateUniversalLoader
      }
}),
```

```
import { TranslateLoader } from '@ngx-translate/core';
import { Observable, of } from 'rxjs';

import * as contentEn from './en.json';
import * as contentRo from './ro.json';

const TRANSLATIONS = {
  en: contentEn,
  ro: contentRo
};

export class TranslateUniversalLoader implements TranslateLoader {
  getTranslation(lang: string): Observable<any> {
    return of(TRANSLATIONS[lang].default);
  }
}

```

It looks like the returned object is added under the default key 
tested this in Angular 7  and also tested dynamic change works just fine

**Comment by psrebrny** [comment]
In Angular 7 this warning wan't resolved

**Comment by vahidvdn** [comment]
@virgil-av How do you import json file as a module? I mean the following:
```
import * as contentEn from './en.json';
```

**Comment by Portekoi** [comment]
any updates?

**Comment by virgil-av** [comment]
> any updates?

Have a look here: https://hackernoon.com/import-json-into-typescript-8d465beded79

From Typescript 2.9
In your `tsconfig.json` file, under compiler options, add these two lines:

```
{
  "compilerOptions": {
    "resolveJsonModule": true,
    "esModuleInterop": true  
  }
}
```

Then you can import json like this:

// Typescript
```
import data from './example.json';
const word = (<any>data).name;
console.log(word); // output 'testing'
```

**Comment by thw0rted** [comment]
Virgil, I think you're on the wrong issue.  This is nothing to do with importing from JSON, it's using dynamic `import()` in place of the SystemJS library.  Maybe you meant to post those comments in a different thread?

(Also, import from JSON is actually an issue with your loader, so it's only going to work if you're running your generated JS through a further build process with e.g. webpack.  I can find you an issue where WhatWG is currently discussing adding native support for it, if you're interested, but there isn't a consensus yet.)

**Comment by craigcosmo** [comment]
do we got official fixes ?

**Comment by wKoza** [comment]
In Angular 8 this warning wan't resolved

**Comment by almartino** [comment]
Please, see the comment by @ermarkar https://github.com/angular/angular/issues/23426#issuecomment-449725566

**Comment by thw0rted** [comment]
Good to know how to hide the warning, @almartino , but the team really should move away from SystemJS imports soon.  It's practically 2020 and we have lots of support for native `import()`.

**Comment by iranicus** [comment]
In Angular 9 this warning wasn't resolved :'(

**Comment by seahindeniz** [comment]
```js
{
    test: /\.js$/,
    parser: { system: false },
}
```

Can anyone care to explain what is the downside of disabling the SystemJS other than freeing the global variable "System"? 

---

Quoting from [Performance](https://github.com/systemjs/systemjs#performance) section
> SystemJS is designed for production modules performance and can load multiple modules in less than a millisecond. Its performance is only around a factor of 1.5 times the performance of native ES modules, as seen in the following performance benchmark, which was run by loading 426 javascript modules (all of @babel/core) on a Macbook pro with fast wifi internet connection. Each test was the average of five page loads in Chrome 80.

**Comment by rahulserver** [comment]
Facing this issue with     `"@angular/core": "^11.2.10",`

**Comment by petebacondarwin** [comment]
@rahulserver - which issue exactly are you facing?

**Comment by rahulserver** [comment]
@petebacondarwin I just imported that angular core library in an ionic project (Which is actually a vue project) because I had to us a barcode scanner module https://ionicframework.com/docs/native/barcode-scanner (which needs angular/core in a vue js project). 
And I get this in console whenever i run the project: 

``` WARNING  Compiled with 2 warnings                                                                                                                                      6:09:30 PM

 warning  in ./node_modules/@angular/core/fesm2015/core.js

Critical dependency: the request of a dependency is an expression

 warning  in ./node_modules/@angular/core/fesm2015/core.js

System.import() is deprecated and will be removed soon. Use import() instead.
For more info visit https://webpack.js.org/guides/code-splitting/

 @ ./node_modules/@ionic-native/barcode-scanner/ngx/index.js 2:0-43 24:16-26
 @ ./node_modules/cache-loader/dist/cjs.js??ref--12-0!./node_modules/babel-loader/lib!./node_modules/cache-loader/dist/cjs.js??ref--0-0!./node_modules/vue-loader-v16/dist??ref--0-1!./src/components/modals/NewPurchaseModal.vue?vue&type=script&lang=js
 @ ./src/components/modals/NewPurchaseModal.vue?vue&type=script&lang=js
 @ ./src/components/modals/NewPurchaseModal.vue
 @ ./node_modules/cache-loader/dist/cjs.js??ref--12-0!./node_modules/babel-loader/lib!./node_modules/cache-loader/dist/cjs.js??ref--0-0!./node_modules/vue-loader-v16/dist??ref--0-1!./src/views/Request.vue?vue&type=script&lang=js
 @ ./src/views/Request.vue?vue&type=script&lang=js
 @ ./src/views/Request.vue
 @ ./src/router/index.js
 @ ./src/main.js
 @ multi (webpack)-dev-server/client?http://192.168.0.100:8080&sockPath=/sockjs-node (webpack)/hot/dev-server.js ./src/main.js

```

**Comment by petebacondarwin** [comment]
So the issue is that you are receiving a warning due to this deprecated `System.import` call in the code.
This will be removed in the next major release (v13 since 12.0.0 is about to ship next week).

**Comment by JoostK** [comment]
`SystemJsNgModuleFactoryLoader` has been removed in #43591.
